persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java

```diff
@@ -269,26 +269,40 @@ public Authenticator getAuthenticator() {
       boolean isTracingRequested,
       long queryStartNanoTime) {
     try {
-      org.apache.cassandra.service.QueryState internalState = Conversion.toInternal(state);
-      org.apache.cassandra.cql3.QueryOptions internalOptions = Conversion.toInternal(options);
-
       return Conversion.toFuture(
           Single.defer(
               () -> {
                 try {
-                  final UUID tracingId =
-                      beginTraceQuery(
-                          cql, internalState, internalOptions, customPayload, isTracingRequested);
+                  org.apache.cassandra.service.QueryState internalState =
+                      Conversion.toInternal(state);
+                  org.apache.cassandra.cql3.QueryOptions internalOptions =
+                      Conversion.toInternal(options);
 
                   checkIsLoggedIn(internalState);
 
                   CQLStatement statement = QueryProcessor.parseStatement(cql, internalState);
+
+                  // Please note that this needs to happen _before_ the beginTraceQuery, because
+                  // when we add bound values to the trace, we rely on the values having been
+                  // re-ordered by the following prepare (if named values were used that is).
                   internalOptions.prepare(statement.getBindVariables());
 
+                  final UUID tracingId =
+                      beginTraceQuery(
+                          cql, internalState, internalOptions, customPayload, isTracingRequested);
                   return processStatement(
-                      statement, state, options, customPayload, queryStartNanoTime, tracingId);
+                      statement,
+                      state,
+                      options,
+                      internalState,
+                      internalOptions,
+                      customPayload,
+                      queryStartNanoTime,
+                      tracingId);
                 } catch (Exception e) {
-                  return Single.error(Conversion.handleException(e));
+                  return Tracing.instance
```

Comment:
It's kind of confusing that we stop the tracing here and exception, but not on the normal path. It does appears this is done in `processStatement`, but at least a comment to that effect would help the reader.

Reply:
Good call, I address this here: https://github.com/stargate/stargate/pull/183/commits/c8203e9d5f518ec53a8890566e1369b422521084